### ISSUE TITLE
Padding zero at end of sequence is unstable if using maskzero 

### ISSUE THREAD

**Issue by puuzll** [Element-Research/rnn#155]
I modified "examples/encoder-decoder-coupling.lua" to test batch training:
```lua
--[[

Example of "coupled" separate encoder and decoder networks, e.g. for sequence-to-sequence networks.

]]--

require 'rnn'

version = 1.2 -- refactored numerical gradient test into unit tests. Added training loop

local opt = {}
opt.learningRate = 0.1
opt.hiddenSize = 6
opt.vocabSize = 5
opt.seqLen = 3 -- length of the encoded sequence
opt.niter = 1000

--[[ Forward coupling: Copy encoder cell and output to decoder LSTM ]]--
local function forwardConnect(encLSTM, decLSTM,seqLen)
   decLSTM.userPrevOutput = nn.rnn.recursiveCopy(decLSTM.userPrevOutput, encLSTM.outputs[seqLen])
   decLSTM.userPrevCell = nn.rnn.recursiveCopy(decLSTM.userPrevCell, encLSTM.cells[seqLen])
end

--[[ Backward coupling: Copy decoder gradients to encoder LSTM ]]--
local function backwardConnect(encLSTM, decLSTM)
   encLSTM.userNextGradCell = nn.rnn.recursiveCopy(encLSTM.userNextGradCell, decLSTM.userGradPrevCell)
   encLSTM.gradPrevOutput = nn.rnn.recursiveCopy(encLSTM.gradPrevOutput, decLSTM.userGradPrevOutput)
end

-- Encoder
local enc = nn.Sequential()
enc:add(nn.LookupTableMaskZero(opt.vocabSize, opt.hiddenSize))
enc:add(nn.SplitTable(1, 2)) -- works for both online and mini-batch mode
local encLSTM = nn.LSTM(opt.hiddenSize, opt.hiddenSize):maskZero(1)
enc:add(nn.Sequencer(encLSTM))
enc:add(nn.SelectTable(-1))

-- Decoder
local dec = nn.Sequential()
dec:add(nn.LookupTableMaskZero(opt.vocabSize, opt.hiddenSize))
dec:add(nn.SplitTable(1, 2)) -- works for both online and mini-batch mode
local decLSTM = nn.LSTM(opt.hiddenSize, opt.hiddenSize):maskZero(1)
dec:add(nn.Sequencer(decLSTM))
dec:add(nn.Sequencer(nn.MaskZero(nn.Linear(opt.hiddenSize, opt.vocabSize),1)))
dec:add(nn.Sequencer(nn.MaskZero(nn.LogSoftMax(),1)))
-- dec = nn.MaskZero(dec,1)

local criterion = nn.SequencerCriterion(nn.MaskZeroCriterion(nn.ClassNLLCriterion(),1))

-- Some example data (batchsize = 2)
local encInSeq1 = torch.Tensor({{1,2,3},{3,2,1}}) 
local decInSeq1 = torch.Tensor({{1,2,3,4},{2,4,3,1}})
local decOutSeq1 = torch.Tensor({{2,3,4,1},{4,3,1,2}})
decOutSeq1 = nn.SplitTable(1, 1):forward(decOutSeq1)
local encInSeq = torch.Tensor({{1,1,1,2,3},{0,0,1,2,3},{0,0,3,2,1}}) 
local decInSeq = torch.Tensor({{1,1,1,1,2,3},{1,2,3,4,0,0},{2,4,3,1,0,0}})
local decOutSeq = torch.Tensor({{1,1,1,2,3,2},{2,3,4,1,0,0},{4,3,1,2,0,0}})
decOutSeq = nn.SplitTable(1, 1):forward(decOutSeq)
print(decOutSeq)


print('encoder:')
for i,module in ipairs(enc:listModules()) do
  print(module)
  break
end
print('decoder:')
for i,module in ipairs(dec:listModules()) do
  print(module)
  break
end
local function train(i,encInSeq, decInSeq,decOutSeq)

   -- Forward pass
   local len = encInSeq:size(2)
   -- print(len)
   local encOut = enc:forward(encInSeq)
   forwardConnect(encLSTM, decLSTM,len)
   local decOut = dec:forward(decInSeq)
   -- print("decout:")
   -- for i = 1,#decOut do
     -- print(decOut[i])
   -- end
   local err = criterion:forward(decOut, decOutSeq)
   -- print(err) 
   print(string.format("Iteration %d ; NLL err = %f ", i, err))

   -- Backward pass

   local gradOutput = criterion:backward(decOut, decOutSeq)
   dec:backward(decInSeq, gradOutput)
   backwardConnect(encLSTM, decLSTM)
   local zeroTensor = torch.Tensor(2):zero()
   enc:backward(encInSeq, zeroTensor)

   dec:updateParameters(opt.learningRate)
   enc:updateParameters(opt.learningRate)
   enc:zeroGradParameters()
   dec:zeroGradParameters()
   dec:forget()
   enc:forget()
   encLSTM:recycle()
   decLSTM:recycle()
end
for i=1,1000 do
  train(i,encInSeq,decInSeq,decOutSeq)
  -- train(i,encInSeq1,decInSeq1,decOutSeq1)
end
```
but err may increased during some itrations ,like:
```
Iteration 89 ; NLL err = 1.579057       
Iteration 90 ; NLL err = 2.345541       
Iteration 91 ; NLL err = 1.374780  
```
It do not apear every time but in high probability.  It may be trivial in such test case but is fatal to train an model on large cropus.
I find this word in "MaskZero.lua":
```lua
-- Zero vectors (i.e. padding) must be at the beginning of the sequence
```
For encoder-decoder model,it is difficult to pad zero at the beginning of decode-sequence.May be I can handle it by myself to record beginning of every sequence in a batch and fix forward/backward at every time step. 
But I want to know what happened in maskzero if pad zero at the end of sequence.

**Comment by nicholas-leonard** [comment]
@puuzll I don't see why we shouldn't be able to mask anywhere in the sequence. I am writing some unit test to make sure it works.

**Comment by nicholas-leonard** [comment]
@puuzll #158 confirms that it is possible to pad anywhere in sequence. If you think it would benefit others, please send a Pull Request with your changes to the encoder-decoder example. Thanks!
